consensus/istanbul/backend/engine.go

```diff
@@ -658,14 +656,12 @@ func (sb *Backend) SetCallBacks(hasBadBlock func(common.Hash) bool,
 
 // StartValidating implements consensus.Istanbul.StartValidating
 func (sb *Backend) StartValidating() error {
-	sb.coreMu.Lock()
-	defer sb.coreMu.Unlock()
-	if sb.coreStarted {
+	if sb.coreStarted.Load().(bool) {
 		return istanbul.ErrStartedEngine
```

Comment:
Hi @YoduYodu, although the switch from a mutex to an atomic avoids the data race that was previously occurring, it has now introduced a race condition. Where before, the first go-routine to call `StartValidating` would execute the function and any subsequent go-routine would return at this line. Now its possible for any amount of go-routines to progress past this check before `coreStarted` is set to true.

I think a mutex is required here and in `StopValidating` to ensure the correct behaviour. I don't see a reason that you couldn't use a mutex and an atomic value together. I'm sure it should be possible with just the mutex but I suspect that it could be tricky to get that working well, given the current state of the code. So using an atomic value and sometimes an additional mutex I think is fine for now.

Reply:
I think using Mu & atomic.Value is a bit verbose & confusing, so I switch back to only using Mu.